design/0002-vscode-extension-wot-cli.org

```diff
@@ -0,0 +1,39 @@
+#+TITLE: VSCode Extension and wot CLI
```

Comment:
Not sure I believe in the VSCode extension, as the persona developing the WoT TD is not a developer but an Ot operator, in my world. Therefore, I really like the fact that the edi[TD]or is hosted for you and you can integrate it into any toolchain onboarding workflow easily...

Reply:
Good point for a brown-field on-boarding scenario. Probably it makes sense to write down scenarios that people have in their mind.

We oftentimes write a couple of TMs and some algorithm that works on on the generated TDs. That would be easier in an integrated environment, with TMs, TD generation and runtime.

Also there are device integrators that write TMs in an back-office for 3rd party devices. They will have a repository of TMs and probably property templates that they will want to access.

Maybe we can get some input from others on the roles of people who will write TMs/TDs.

Reply:
So regarding integration into a workflow:

We need to understand what people expect from such TD editors and tools. From my point of view, EdiTDor is now a Web UI so it cannot be integrated into a workflow per se but its specific features can be or it can be used (not integrated) as a tool in a workflow (like Postman). Integratabiltiy is what [TD Playground](https://github.com/thingweb/thingweb-playground) does where different packages are offered as a Node.js dependency, [CLI](https://github.com/thingweb/thingweb-playground/tree/master/packages/cli) or integrated in a [Web UI](http://plugfest.thingweb.io/playground/). 

Anyways I think a separate discussion/issue is needed for this point :)

Reply:
@hadjian I think it would be worth to invite for a short call to layout what we already have and do (and open source), as:
- TD Playground mentioned by @egekorkan above, where @edbingo did a full rework of the cli td validator under the lead of @mkovatsc + all the other stuff on TD Playground @egekorkan and others are doing anyways (also with @sebastiankb towards w3c standardization)
- the ediTDor of course
- then beginning of this year @ektrah open sourced a type script based ontology viewer, running in browser with [rdf-toolkit](https://github.com/ektrah/rdf-toolkit)
- ... maybe more

All this has the potential to be used together / integrate / run side-by-side ... whatever we can imagine and makes sense

To make a statement here for the VSC extension per se, I'm not to deep in the daily usage (and pain) of TM/TD handling. I'd need here the input from @fennibay , @mkovatsc and further colleagues doing this already.

But anyhow, my two cents on this: VSC extension's main benefit is to do developers life easier, where you code and have TM/TD side-by-side and would enjoy a great support handling them by an extension. And if this applies, and let's assume an extension is created, don't be surprised for receiving more features to be delivered. (as e.g. TM/TD object based comparison of two TM/TD's (diffing))

Reply:
Great pointers. I will invite for a one to one meeting and put a summary here. And your mention of this side-by-side workflow is what I am interested in.

Reply:
I agree with @barnstee. 
Having the ediTDor integrated into our Thing Model repository is most beneficial for us. In this way, we have an easy to use single source of truth for Thing Models.

Reply:
VSC Extension would be helpful for development cases, I agree.

What we'd like even more is a reusable version of editdor that we can integrate in our engineering software, e.g. as a web component. We can also talk to Ozgur Gulderen about this, who is working on such an engineering software.

Reply:
I join the comment train :). Internally we use the [td-code](https://marketplace.visualstudio.com/items?itemName=arces-wot.td-code) and we usually edit what I would call [Partial TDs](https://www.w3.org/TR/wot-architecture11/#dfn-partial-td) alongside the code (see an old experiment [here](https://github.com/vaimee/WoTSimFarm/tree/master/backend) - it is also a good example of how you can use node-wot as a runtime+cli). I would say that this workflow is now a little bit outdated cause it predates the introduction of Thing Models, but in general, I expect that to stay and I have a couple of inputs about what I would like to see in a full-fledge VSCode extension:
- Intellisense for Json-ld --> meeting understanding terms imported in contexts (vscode currently supports only minimal features meaning it uses the JSON extension on json-ld files see https://github.com/microsoft/vscode/issues/95013 and https://github.com/microsoft/vscode/issues/56334)
- Intellisense for ThingModels --> meaning I can get autocomplete and validation from the linked TMs
- I usually don't like visual tooling when coding, but for some people, it might be helpful to have a rendered preview of the TM/PartialTD with the linked TMs (it could accessible by the usual ctrl+shift+v shortcut)

Reply:
Perfect. It seems very clear that the ediTDor approach is working, so we should start there and improve. Also, there seem to be quite some usable components available that I didn't know of. @egekorkan is giving me a tour today. I will adjust the proposal and get back to all of you. Good thing we are doing this open source. The feedback is pure gold ;-)